unused.1
.\" Copyright 2024 Duncan McIntosh
.\" SPDX-License-Identifier: GPL-3.0-only
.Dd 2024-01-17
.Dt UNUSED 1
.Os unused
.Sh NAME
.Nm unused
.Nd suggests unused functions in an executable
.Sh SYNPOSIS
.Nm unused
.Ar file
.Sh DESCRIPTION
.Nm unused
finds unused functions within a native executable by disassembling and
parsing the result.  The only command-line option accepted is the
executable to examine.
.Pp
The addresses will be printed to standard output when complete. Each
line corresponds to one function, and there are two tab-separated
fields: the address within the executable, and the name of the
function.
.Sh BUGS
.Nm unused
often finds false positives or false negatives. As such, a function
that
.Nm unused
complains about may actually be used, or unused functions may be
missed; it is more useful as a hint than a hard lint or requirement.
.Pp
All unused functions can be reported, even if they are used in other
build configurations of your program. For example,
.Fn do_thing
may only be referred to if
.Va HAVE_THING_DOER
is defined. In this case, unused may complain if your program was
comopiled without that flag, when in rainbow-land it would not. (Note
that
.Nm unused
is completely unaware of source code.)
.Sh IMPLEMENTATION NOTES
There are two backends: bddisasm and bfd. There is currently no way to
determine which build you have, and they can produce somewhat
different results. Although both should give a general idea, you may
want to run both to be certain.
.Sh EXIT STATUS
.Nm unused
exits with status 0 if it was successful, and 1 if any error occurred
(executable did not exist, read failure, invalid file...)
.Sh EXAMPLES
To determine if the shell has any unused functions, try:
.Dl unused /bin/sh
.Sh AUTHORS
.An Duncan McIntosh Aq Mt duncan82013@live.ca


ChangeLog
2024-01-17  Duncan McIntosh  <duncan82013@live.ca>
	Remove uu_dict_lookup, as it's (ironically) unused.

	Found with unused itself!
	* dict.c: Remove implementation.
	* dict.h: Remove prototype.

2024-01-17  Duncan McIntosh  <duncan82013@live.ca>
	Add some documentation.

	This includes an mdoc-based manual page and README, to try and make
	the project somewhat more discoverable.

	* unused.1: Simple manual page for the program.
	* README.md: Introduction to the project.

2024-01-17  Duncan McIntosh  <duncan82013@live.ca>
	Use updated init_diassemble_info prototype.

	This presumably changed with a recent binutils release, so
	keep using the same code but add a thunk for compatibility.

	* backends/libopcodes.c: Add a thunk for fprintf_styled_ftype.

2021-03-22  thatlittlegit  <duncan82013@live.ca>
	Adds a BdDisasm backend. This is faster, but seems to give slightly
	different results; it seems to be fairly close, though.

	Committed 2021-04-09.

	* backends/libbddisasm.c: Add a bddisasm backend.
	* meson.build, meson_options.txt: Add a bddisasm option.

2021-03-22  thatlittlegit  <duncan82013@live.ca>
	This adds an option for which disassembler to use. The default is, and
	will likely remain, libopcodes.

	* meson.build: Only build libopcodes support if told.
	* meson_options.txt: Add an option to select the disassembler.

2021-03-22  thatlittlegit  <duncan82013@live.ca>
	* meson.build: Reformat.

2021-03-22  thatlittlegit  <duncan82013@live.ca>
	Clean up include statements with include-what-you-use.

	* backend.h: Forward-declare uu_dict.
	* backend/libopcodes.c: Include string.h and bfd.h instead of relying
	on other files.
	* dict.h, dict.h: Remove include of <stdint.h>.

2021-03-22  thatlittlegit  <duncan82013@live.ca>
	Separate the libopcodes backend into a separate file.

	Now, more backends can be fairly easily added. I'll need to do some
	work with the Meson file first, but that's just making it look better.
	Then, just implementing uu_backend creates an entire new backend with
	minimal work.

	* backend/libopcodes.c, unused.c: Separate the libopcodes backend.
	* unused.c: Use uu_backend instead of process_section.
	* backend.h: Add a header for uu_backend.
	* meson.build: Also build backend/libopcodes.c.

2021-03-22  thatlittlegit  <duncan82013@live.ca>
	Only remove the symbol from the table when we see it.

	We wouldn't reset fprintf_data.value, and therefore we would call
	uu_dict_remove on every instruction. I think this shaves a second, but
	I don't think it's that much (probably more chance.)

	* unused.c: Reset fprintf_data.value between iterations.

2021-03-22  thatlittlegit  <duncan82013@live.ca>
	Only buffer the disassembly if we're actually going to look at it.

	We used to need to do all of this, but now with fake_fprintf there's
	no need to.

	* unused.c: Only buffer disassembly if needed.

2021-03-22  thatlittlegit  <duncan82013@live.ca>
	Reorganize fake_fprintf a little. No real change, found when trying to
	optimize (unsuccessfully).

	* unused.c: Reorganize fake_fprintf

2021-03-21  thatlittlegit  <duncan82013@live.ca>
	Only fprintf to debug if we're actually in debug mode.

	It turns out this saves 7s with libxul!

	* debug.h: Add a debugging function header.
	* debug.c: Add a waste of space.
	* unused.c: Use debug() instead of fprintf(debug).
	* meson.build: Build debug.c.

2021-03-21  thatlittlegit  <duncan82013@live.ca>
	Use a dictionary to keep track of functions instead of lists.

	This means that we no longer have to look through the list of symbols
	so much, and thus can be very fast. libxul takes 17s (!) now!

	* dict.h: Add a dictionary header.
	* dict.c: Add a basic dictionary implementation.
	* unused.c: Use a dictionary instead of arrays to store symbols.
	* meson.build: Build dict.c.

2021-03-21  thatlittlegit  <duncan82013@live.ca>
	Use a custom function to replace fprintf unless we're in debug mode.

	fake_fprintf() avoids running fprintf, and it is a lot faster! I found
	before I could run this on libxul.so as a test, and it not once
	started printing anything out. Now? Thirty seconds or so and we're
	printing. Incredible.

	* unused.c: Use a custom function to replace fprintf.

2021-03-21  thatlittlegit  <duncan82013@live.ca>
	Allows compilation with C++. This isn't very well supported by Meson
	at this point, but it does mean that we meet higher type-safety
	standards. I mean, we cast everything, but oh well.

	* unused.c: Allow compilation with C++ by adding casts on
	reallocarray.

2021-03-19  thatlittlegit  <duncan82013@live.ca>
	Fixes the debug FILE* not having correct permissions, or a potential
	segfault if /dev/null can't be opened for some reason.

	* unused.c: Fix the debug stream's permissions and error handling.

2021-03-19  thatlittlegit  <duncan82013@live.ca>
	Fixes some memory leaks in unused.c.

	* unused.c: Fix some memory leaks.

2021-03-18  thatlittlegit  <duncan82013@live.ca>
	Migrate to ANSI C. Why? Why not?

	One inspiration for this was the Hundred Rabbits <https://100r.co>,
	who have written plenty of utilities in ANSI C.

	It's surprisingly not that bad, but I don't know why I do this to
	myself.

	* unused.c: Get rid of sanity.
	* unused.c: Define _DEFAULT_SOURCE and _POSIX_C_SOURCE to get access
	to various functions that aren't available in C89 by default.
	* meson.build: Use c89 as the default C standard version.

2021-03-18  thatlittlegit  <duncan82013@live.ca>
	This adds disassembly so then we can actually see what symbols are
	actually being used. Frankly, it's surprisingly close to done at this
	point! :D

	* unused.c: Use libopcodes to do disassembly.
	* meson.build: Add libopcodes as a dependency.

2021-03-18  thatlittlegit  <duncan82013@live.ca>

	* .clang-format: Add basic clang-format file with some alright formatting.
	* meson.build: Add initial Meson build file.
	* unused.c: Add basic C skeleton that prints all the functions in a file.



unused.c
/* unused.c: main file for unused
 *
 * Copyright (C) 2021 thatlittlegit
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, under version 3 of the License,
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
#define _DEFAULT_SOURCE

#include "backend.h"
#include "debug.h"
#include "dict.h"
#include <assert.h>
#include <bfd.h>
#include <errno.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

const char* progname = "unused";

struct options {
	const char* filename;
	int debug;
};

struct definition {
	struct definition* next;
	const char* name;
	unsigned id;
};

static int
parse_options(int argc, char** argv, struct options* opts)
{
	int c;
	opterr = 0;
	while ((c = getopt(argc, argv, "d")) > 0)
		switch (c) {
		case 'd':
			opts->debug = 1;
			break;
		default:
			fprintf(stderr, "%s: unknown argument '%c'\n", progname,
				c);
			return -1;
		}

	opts->filename = argv[optind];
	return 0;
}

static int
get_symbols(bfd* bfd,
	    struct bfd_symbol*** static_syms,
	    unsigned* static_len,
	    struct bfd_symbol*** dynamic_syms,
	    unsigned* dynamic_len)
{
	long static_size;
	long dynamic_size;

	*static_syms = NULL;
	*dynamic_syms = NULL;

	static_size = bfd_get_symtab_upper_bound(bfd);
	if (static_size < 0)
		goto fail;

	dynamic_size = bfd_get_dynamic_symtab_upper_bound(bfd);
	if (dynamic_size < 0)
		goto fail;

	*static_syms = (struct bfd_symbol**)malloc(static_size);
	if (static_size > 0 && *static_syms == NULL)
		goto fail;

	/* why is the function named this?! */
	*static_len = bfd_canonicalize_symtab(bfd, *static_syms);

	*dynamic_syms = (struct bfd_symbol**)malloc(dynamic_size);
	if (dynamic_size > 0 && *dynamic_syms == NULL)
		goto fail;

	*dynamic_len = bfd_canonicalize_dynamic_symtab(bfd, *dynamic_syms);

	return 0;

fail:
	free(*static_syms);
	free(*dynamic_syms);

	return -1;
}

static void
process_symbol(struct bfd_symbol* sym, void* data)
{
	bfd_vma location;
	(void)data;

	location = sym->value + sym->section->vma;
	fprintf(stdout, "%lx\t%s\n", location, sym->name);
}

int
main(int argc, char** argv)
{
	int status;
	struct options opts;
	FILE* input;

	bfd* bfd;
	struct bfd_symbol** static_symbols;
	unsigned static_len;
	struct bfd_symbol** dynamic_symbols;
	unsigned dynamic_len;

	struct uu_dict* table;

	unsigned i;

	status = EXIT_SUCCESS;
	input = stdin;

	opts.filename = NULL;
	opts.debug = 0;

	assert(bfd_init() == BFD_INIT_MAGIC);

	if (argc >= 1)
		progname = argv[0];

	if (parse_options(argc, argv, &opts) < 0) {
		status = EXIT_FAILURE;
		goto cleanup_none;
	}

	if (opts.filename != NULL)
		input = fopen(opts.filename, "r");

	if (input == NULL) {
		fprintf(stderr, "%s: failed to open input file '%s': %s\n",
			progname, opts.filename, strerror(errno));
		status = EXIT_FAILURE;
		goto cleanup_none;
	}

	if (opts.filename == NULL)
		opts.filename = "(stdin)";

	if (opts.debug)
		debug_stream = stderr;
	else
		debug_stream = NULL;

	bfd = bfd_openstreamr(opts.filename, NULL, input);
	if (bfd == NULL) {
		fprintf(stderr, "%s: failed to read input file '%s': %s\n",
			progname, opts.filename, bfd_errmsg(bfd_get_error()));
		status = EXIT_FAILURE;
		goto cleanup_stream;
	}

	if (!bfd_check_format(bfd, bfd_object)) {
		fprintf(stderr, "%s: failed to get type of '%s': %s\n", progname,
			opts.filename, bfd_errmsg(bfd_get_error()));
		status = EXIT_FAILURE;
		goto cleanup_bfd;
	}

	if (get_symbols(bfd, &static_symbols, &static_len, &dynamic_symbols,
			&dynamic_len)
	    < 0) {
		fprintf(stderr, "%s: failed to get symbols in '%s': %s\n",
			progname, opts.filename, bfd_errmsg(bfd_get_error()));
		status = EXIT_FAILURE;
		goto cleanup_bfd;
	}

	table = uu_dict_new();

	for (i = 0; i < static_len + dynamic_len; i++) {
		struct bfd_symbol* sym;

		if (i < static_len)
			sym = static_symbols[i];
		else
			sym = dynamic_symbols[i - static_len];

		if (!(sym->section->flags & SEC_CODE))
			continue;
		if (sym->value == 0)
			continue;

		uu_dict_add(table, sym->value + sym->section->vma, sym);
	}

	bfd_map_over_sections(
	    bfd, (void (*)(struct bfd*, asection*, void*))uu_backend, table);
	uu_dict_for_each(table, (void (*)(void*, void*))process_symbol, NULL);

	uu_dict_free(table);
	free(static_symbols);
	free(dynamic_symbols);

cleanup_bfd:
	bfd_close(bfd);
	input = NULL;
cleanup_stream:
	if (input)
		fclose(input);
cleanup_none:
	return status;
}
